edb/server/protocol/ai_ext.py

```diff
@@ -363,22 +363,51 @@ async def _ext_ai_fetch_active_models(
     return result
 
 
-_EXT_AI_ADVISORY_LOCK = b"3987734540"
+_EXT_AI_ADVISORY_LOCK_LOCK = b"3987734540"
+_EXT_AI_ADVISORY_LOCK = b"3987734541"
 
 
 async def _ext_ai_lock(
+    tenant: srv_tenant.Tenant,
     pgconn: pgcon.PGConnection,
 ) -> bool:
-    b = await pgconn.sql_fetch_val(
-        b"SELECT pg_try_advisory_lock(" + _EXT_AI_ADVISORY_LOCK + b")")
-    return b == b'\x01'
+    async def _try_xact_lock(lock: bytes) -> bool:
+        b = await pgconn.sql_fetch_val(
+            b"SELECT pg_try_advisory_xact_lock(" + lock + b")")
+        return b == b'\x01'
+
+    # We use transaction-level advisory locks to ensure releasing
+    await pgconn.sql_execute(b"START TRANSACTION")
+    try:
+        # Acquire the outer lock to operate on the real lock
+        if await _try_xact_lock(_EXT_AI_ADVISORY_LOCK_LOCK):
+            # If success, acquire the real lock in the current session
+            # temporarily to make sure it's not locked by system session
+            sp = b"ext_ai_lock_sp"
+            await pgconn.sql_execute(b"SAVEPOINT " + sp)
+            if await _try_xact_lock(_EXT_AI_ADVISORY_LOCK):
+                # If success, move the lock to the system session
+                await pgconn.sql_execute(b"ROLLBACK TO SAVEPOINT " + sp)
+                async with tenant.use_sys_pgcon() as syscon:
```

Comment:
What is the try-lock for?
And couldn't it get taken by another connection after the ROLLBACK but before the lock?
Is that a problem?

Reply:
> What is the try-lock for?

It checks if the lock is being held by another worker (through sys_pgcon). We cannot acquire this lock directly in sys_pgcon because, say if another worker is already holding this lock through sys_pgcon, re-acquiring the lock will not block (or fail with the try_* method) from the same sys_pgcon session.

> And couldn't it get taken by another connection after the ROLLBACK but before the lock?

That's guaranteed not to happen by the outter/first lock, unless other code misused the same advisory lock ID.